src/dos/dos_code_page.cpp

```diff
@@ -0,0 +1,2223 @@
+			}
+		}
+
+		bool should_skip = false;
+
+		// Check if code page is valid
+		if (code_page == 0) {
+			WarnInvalidCodePage(header->struct_offset,
+			                    header->StructName,
+			                    code_page);
+			should_skip = true;
+		}
+
+		// Check if device name is valid
+		const bool is_device_valid = IsDeviceNameValid(*header);
+		if (!is_device_valid) {
+			WarnInvalidDeviceName(header->struct_offset,
+			                      header->StructName);
+			should_skip = true;
+		}
+
+		const auto cpih_offset      = adapt_cpih(header->cpih_offset);
+		const auto next_cpeh_offset = adapt_cpeh(header->next_cpeh_offset);
+
+		const uint32_t base_offset = IsFormatWinNt() ? header->struct_offset
+		                                             : 0;
+		const uint32_t font_offset = base_offset + cpih_offset;
+		if (font_offset < GetCurrentOffset()) {
+			// Going backwards is not allowed, skip this font
+			WarnOffsetPointingBackwards(header->struct_offset,
+			                            header->StructName);
+			should_skip = true;
+		}
+
+		if (!should_skip) {
+			StoreFontType(*header, result);
+		}
+
+		const bool is_screen = IsScreenCodePage(*header);
+		const bool is_ega    = IsEgaCodePage(*header);
+		const bool is_lcd    = IsLcdCodePage(*header);
+
+		if (!should_skip && is_screen &&
+		    IsCodePageMatching(*header, code_page_filter)) {
+			const auto font = GetScreenFont(font_offset);
+			if (check_screen_font(font)) {
+				if (is_ega && !result_ega.contains(code_page)) {
+					result_ega[code_page] = *font;
+				}
+				if (is_lcd && !result_lcd.contains(code_page)) {
+					result_lcd[code_page] = *font;
+				}
+				if (is_screen && !is_ega && !is_lcd &&
+				    !result_other.contains(code_page)) {
+					result_other[code_page] = *font;
+				}
+			}
+		}
+
+		// If it was the last structure - terminate
+		if (idx + 1 == num_code_pages) {
+			break;
+		}
+
+		if (!IsFormatWinNt()) {
+			// Some files (mostly unofficial ones, but also the
+			// '4208.CPI' from MS-DOS 6.22) have too high number of
+			// code pages in the 'FontInfoHeader', but they mark the
+			// end of usable content with a special offset value.
+			if (next_cpeh_offset == GetContentSize() ||
+			    DummyOffsetValues.contains(next_cpeh_offset)) {
+				break;
+			}
+		}
+
+		// If not the last code page, seek to the next one
+		const uint32_t next_offset = base_offset + next_cpeh_offset;
+		if (next_offset < GetCurrentOffset()) {
+			// Going backwards is not allowed; terminate to prevent
+			// a possible infinite loop
+			WarnOffsetPointingBackwards(header->struct_offset,
+			                            header->StructName);
+			// At least one CPI file ('4208.CPI' from MS-DOS 6.22)
+			// is known to fail this check just before the last
+			// font (the offset is 0).
+			// Since this is a printer font (not useful right now),
+			// there was no in-depth investigation.
+			break;
+		}
+		SetCurrentOffset(next_offset);
+		++idx;
+	}
```

Comment:
Can you make this method less huge? Break it up into logical sub-methods?

Reply:
OK, extracted some parts to separate methods.